File: composition/java/Rectangle.java
public class Rectangle implements Rectangular {
  private int width, height;
  private int x, y;

  public Rectangle(int width, int height, int x, int y) {
    this.width = width;
    this.height = height;
    this.x = x;
    this.y = y;
  }

  public int getWidth() {
    return this.width;
  }

  public Rectangular setWidth(int size) {
    this.width = size;
    return this;
  }

  public int getHeight() {
    return this.height;
  }

  public Rectangular setHeight(int size) {
    this.height = size;
    return this;
  }

  public int getX() {
    return this.x;
  }

  public Rectangular setX(int x) {
    this.x = x;
    return this;
  }

  public int getY() {
    return this.y;
  }

  public Rectangular setY(int y) {
    this.y = y;
    return this;
  }

  public int perimeter() {
    return width * 2 + height * 2;
  }

  public int area() {
    return width * height;
  }

  public Rectangular augmentWidthBy(int size) {
    this.width += size;
    return this;
  }

  public Rectangular augmentHeightBy(int size) {
    this.height += size;
    return this;
  }

  public Rectangular rotate() {
    if(height != width) {
      int tmpWidth = width;
      this.width = this.height;
      this.height = tmpWidth;
    }
    return this;
  }

  public String toString() {
    return "Rectangle:\n\twidth: " + getWidth() + "\theight: " + getHeight() +
    "\n\tx: " + getX() + "\t\ty: " + getY() +
    "\n\tperimeter: " + perimeter() + "\tarea: " + area();
  }
}

File: composition/ruby/rectangle.rb
class Rectangle
  attr_accessor :width, :height, :x, :y

  def initialize(width, height, x, y)
    @width, @height = width, height
    @x, @y = x, y
  end

  def perimeter
    width * 2 + height * 2
  end

  def area
    width * height
  end

  def augment_width_by(size)
    Rectangle.new width + size, height, x, y
  end

  def augment_width_by!(size)
    @width += size
    self
  end

  def augment_height_by(size)
    Rectangle.new width, height + size, x, y
  end

  def augment_height_by!(size)
    @height += size
    self
  end

  def rotate
    Rectangle.new height, width, x, y
  end

  def rotate!
    unless width == height
      old_width = width
      @width = height
      @height = old_width
    end
    self
  end

  def to_s
    string = <<-EOF
Rectangle:
width: #{width}\theight: #{height}
x: #{x}\t\ty: #{y}
perimeter: #{perimeter}\tarea: #{area}
    EOF
  end

end